### DISCUSSION TITLE
ddetailer script does not cooperate with Dynamic Prompts

### DISCUSSION THREAD

**Discussion by 0woon92** [adieyal/sd-dynamic-prompts#313] [Q&A]
Hello adieyal. First of all, I'd like to mention that I'm using a translator to communicate, so there might be some grammatical issues in my message. Please bear with me.

I'm a regular user of your dynamic prompts, and I think your script is fantastic. However, the reason I'm here to ask a question today is slightly different. Are you familiar with a script called ddetailer? To put it simply, this script performs an additional i2i function during the t2img generation process (here's the script's URL: https://github.com/dustysys/ddetailer). I'm asking you about this because you've previously worked on ensuring that the wildcard feature is well-implemented in various scripts.

The problem is that this script hasn't been updated for a long time, and it has some issues. Notably, the wildcard and random selection (ex, {A|B}) features provided by dynamic prompts are not applied. I have some basic Python coding knowledge, so I looked at the script, and I found that the ddetailer script takes the prompt input on the webui as-is when generating an image (without applying the features of dynamic prompts).

So, in the first stage of generation, the transformed prompts from dynamic prompts are output correctly, but in the second i2i stage, the untransformed original prompt is output as-is (for example, if the "\_\_temp\_\_" wildcard was applied to a random prompt in the first stage, in the second stage, the prompt will be applied as "\_\_temp\_\_" as written in the original text).

Would it be possible to ask if there's a way to receive the transformed prompts from dynamic prompts in another script during the t2i process?

The situation is quite frustrating because the original author isn't updating the script. I'm not sure if it's appropriate to ask you this question, but I need help, so I'm leaving this message. I'm looking forward to your positive response. Thank you.

**Comment by adieyal** [comment]
I don't have the time at the moment to look into this in detail, however, here are some thoughts they may be useful.

1. DP runs as an extension and only modifies `p.all_prompts` and `p.all_negative_prompts` ([src](https://github.com/adieyal/sd-dynamic-prompts/blob/main/sd_dynamic_prompts/dynamic_prompting.py#L446-L447)]. 
2. Any downstream scripts should then use those variables.
3.  ddetailer only uses the main prompt ([src](https://github.com/dustysys/ddetailer/blob/master/scripts/ddetailer.py#L198-L199)) - so you might try to change that to all_prompts and see if it works

Otherwise I can take a look once I'm back at my computer.

**Reply by 0woon92** [reply]
Oh, adieyal. First of all, thank you for responding to my question.
I am already aware that the result of running DP is reflected in p.all_prompts.
So, I've tried various things to utilize p.all_prompts as you suggested, but in ddetailer, p.all_prompts has a None value.

What I found out through various tests is that when you click to create an image in the WEBUI, ddetailer and DP run simultaneously (or ddetailer takes precedence), so the prompt inputted into ddetailer has a higher priority for the original prompt rather than the DP processed prompt.

Original Prompt(O.P) -> &nbsp;&nbsp;  ddetailer(input O.P)  &nbsp;&nbsp; ->  &nbsp;&nbsp;&nbsp;  waiting 1st step    &nbsp;&nbsp;&nbsp;-> DP img + O.P ddetailer -> output
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;└-> DP(transfer prompt) -> DP making 1st step -┘

This is the rough flow I've figured out. If you come up with any additional ideas based on the information I provided, I would appreciate it if you could suggest them again. Also, if other users reading this post have the same concerns as me, it would be great to share opinions. Thank you once again, adieyal!

**Comment by adieyal** [comment]
I just tested out the ddetail script. I don't understand what you mean by 1st step and 2nd step. In image2image, if I have an image of a person and I have a prompt such as `A {happy|sad} man staring into the distance`. The prompt is processed by DP and then applied to the mask correctly. Is that what you mean by 1st step? If so, which part is not working?

**Comment by 0woon92** [comment]
Hello, adieyal! I apologize for the delay in my response, as I have been quite busy and only now had the chance to see your reply. I would like to start by expressing my gratitude for conducting the test for my request.

The issue I'm experiencing is related to the use of the wildcard function. To explain the difference between the first and second steps, the first step is where we typically input the prompt used for generating an image, while the second step involves ddetailer receiving that prompt and using it for another round of img2img processing.

Upon examining the code for ddetailer, it takes the prompt from the first step and uses it for the second step. This is where the problem arises.

In the first step, the prompt works with the wildcard applied. However, the prompt obtained by ddetailer from the first step (which is used in the second step) does not have the wildcard function applied!

Here's a clearer example:
The prompt I input: man, people, \_\_wildcard\_\_
The prompt used in the first step: man, people, (wildcard's text)
The prompt used in the second step (ddetailer): man, people, \_\_wildcard\_\_

So, I want to use the prompt "man, people, (wildcard's text)" in both the first and second steps, but currently, ddetailer is unable to receive the prompt with the wildcard applied from the first step for use in the second step, which is the problem I'm encountering.

Did your test include the part I mentioned? Checking through metadata may be less accurate than expected.

The way I tested it was by actually outputting the prompt content received by ddetailer in the code, so I could directly see the prompt being used.

I would really appreciate it if you could help me with my issue once again. Wishing you a happy day! :)
